8263018: Improve API for lifecycle of native resources

mcimadamore opened the pull request:
This patch brings the API improvements described in [1]. Note that this is a significant reshuffle of the API, and might have non-trivial source compatibility impact; the main moves are listed below:

* MemorySegment is no longer AutoCloseable (ResourceScope is, and a MemorySegment *has a* ResourceScope)

* Resources created without an explicit scope (e.g. `MemorySegment::allocateNative(long)`) get a *default* scope, which is non-closeable and GC managed. In other words, users who do not bother with resource scopes, will get same functionalities (lifecycle-wise) that they get with the ByteBuffer API

* Support for custom allocators is added via the `SegmentAllocator` interface; now all API points which need to perform some allocation will accept an explicit `SegmentAllocator` parameter. Where SegmentAllocator-less overloads are provided, the assumed semantics is that of `MemorySegment::allocateNative(long, long)` which means the returned segments will be associated with the *default scope* and will **not** be closeable.

* The method handles generated by the linker will accept an additional leading parameter of type `SegmentAllocator` whenever the native function returns a struct by value. There is an overload which allows to statically bind an allocator at MH creation time.

* The NativeScope abstraction has been removed. This follows the observation that with `SegmentAllocator::arenaBounded/Unbounded` clients can already get pretty close to the functionalities provided by a NativeScope (that said, at least initially, it is likely that jextract will emit a NativeScope abstraction in the generated code, to minimize compatibility impact).

* As discussed in [1], the new API has a way to prevent a resource scope from being closed when operating on a resource associated with that scope; this is `ResourceScope::acquire` which returns a so called *resource scope handle* (an AutoCloseable).

* Access modes are gone. We only kept read-only views (as they are needed to support mapped memory). Non-closeable segments can be obtained by using the acquire API.

* Several methods in MemorySegment are also gone; e.g. all methods related to ownership changes (`withOwnerThread`, `share`) as well as some of the predicate methdos (e.g. `ownerThread`, which is now available through the segment's scope). The overall philosophy is that a scope is assigned to a segment on creation; the scope contains details about how the segment is to be accessed, and these details cannot be altered after the fact.

* `MemoryAddress::asSegmentRestricted` now takes an optional ResourceScope parameter - the scope to be associated with the returned (unsafe) segment. There is also an overload that doesn't take a ResourceScope, in which case the *global scope* (singleton, non-closeable scope) will be used. A similar argument applies to `VaList.ofAddressRestricted`.

* The linker will now accept a scope for the returned upcall stub segment - if no scope is provided, a default one (GC-managed, non-closeable) will be created instead. Same for `VaList::make`.

Overall, I think this patch makes the API cleaner by clarifying the role between lifcycles (e.g. resource scopes) and resources (segments, va lists, etc.) , w/o making clients much more verbose. We also did some internal validation (thanks Chris) to make sure that async byte buffer operation could be adjusted using the *acquire* mechanism. There are some minor outstanding issues which will need to be resolved (as part of this PR, or as a followup) - listed below:

* should the default scope accept custom cleanup actions? Since this scope is created internally using our cleaner factory, I think there's a possibility that user-defined cleanup action might stall the cleaner forever.

* should ResourceScope have a predicate to see if it's a default scope? How should it be called? In an earlier iteration I had `isCloseable` which I don't think does a good job (as a scope can also not be closeable for different reasons).

* Are we ok with the ResourceScope::acquire/ResourceScope.Handle names? Also note that the latter is a simple AutoCloseable, but we need a subclass because the main AutoCloseable::close throws Throwable. Which is unfortunate.

* What about NativeScope? Are we ok with *not* having it? Note that, without NativeScope, code like:

```
try (NativeScope scope = NativeScope.ofUnbounded()) {
   ...
}
```

becomes:

```
try (ResourceScope scope = ResourceScope.ofConfined()) {
   SegmentAllocator allocator = SegmentAllocator.arenaUnbounded(scope);
   ...
}
```

E.g. one extra line in the user code. I believe this is not a huge deal, in the sense that in all our example (most of which are jextract based), the body of the try with resources if pretty biggie in comparison. That said, I'm open to reintroduce NativeScope - although probably in a different form - e.g. by having a sub-interface of SegmentAllocator called BoundedAllocator, which is essentially an allocator that *has a* scope. But we can also add this later depending on our experience with using the API.

That's all for now. Feedback welcome!

[1] - https://inside.java/2021/01/25/memory-access-pulling-all-the-threads/

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263018](https://bugs.openjdk.java.net/browse/JDK-8263018): Improve API for lifecycle of native resources


### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - Committer) ⚠️ Review applies to d02e86f9fad8ad4ed17d84485e9fd198aa9eea73


### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/466/head:pull/466`
`$ git checkout pull/466`

To update a local copy of the PR:
`$ git checkout pull/466`
`$ git pull https://git.openjdk.java.net/panama-foreign pull/466/head`

mcimadamore commented:
Javadoc available here: http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v3/javadoc/jdk/incubator/foreign/package-summary.html

JornVernee commented on test/micro/org/openjdk/bench/jdk/incubator/foreign/Upcalls.java:
```diff
@@ -26,6 +26,7 @@
 import jdk.incubator.foreign.LibraryLookup;
 import jdk.incubator.foreign.MemoryAddress;
 import jdk.incubator.foreign.CLinker;
+import jdk.incubator.foreign.ResourceScope;
```
Spurious import?

JornVernee commented on test/micro/org/openjdk/bench/jdk/incubator/foreign/libCallOverheadJNI.c:
```diff
@@ -37,3 +33,29 @@ JNIEXPORT jint JNICALL Java_org_openjdk_bench_jdk_incubator_foreign_CallOverhead
   (JNIEnv *env, jclass cls, jint x) {
     return identity(x);
 }
+
+JNIEXPORT jobject JNICALL Java_org_openjdk_bench_jdk_incubator_foreign_CallOverhead_identityStruct
+  (JNIEnv *env, jclass cls, jobject pointBuf) {
+    Point *pInput = (Point*)(*env)->GetDirectBufferAddress(env, pointBuf);
+    Point p = identity_struct(*pInput);
+    Point *ret = (Point*)malloc(sizeof(Point));
+    ret->x = p.x;
+    ret->y = p.y;
+    return (jobject)(*env)->NewDirectByteBuffer(env, ret, sizeof(Point));
+}
+
+JNIEXPORT jobject JNICALL Java_org_openjdk_bench_jdk_incubator_foreign_CallOverhead_identityStructAlloc
+  (JNIEnv *env, jclass cls, jobject pointBuf, jobject resBuf) {
+    Point *pInput = (Point*)(*env)->GetDirectBufferAddress(env, pointBuf);
+    Point p = identity_struct(*pInput);
+    Point *pOutput = (Point*)(*env)->GetDirectBufferAddress(env, resBuf);
+    pOutput->x = p.x;
+    pOutput->y = p.y;
+    return resBuf;
+}
+
+JNIEXPORT void JNICALL Java_org_openjdk_bench_jdk_incubator_foreign_CallOverhead_freeStruct
+ (JNIEnv *env, jclass cls, jobject pointBuf) {
+    void *buf = (void*)(*env)->GetDirectBufferAddress(env, pointBuf);
+    free(buf);
+}
```
These seem unused?

mcimadamore replied:
whoops!

JornVernee reviewed (commented):
Amazing work! (it was pleasure reading this) I think this is a great step forward. MemorySegment is now a much more minimal API, and lifetime concerns are nicely contained in ResourceScope.

I tried to be very thorough with the review, since I was not that involved with the design this time around (working on upcall specialization at the same time). Also just asking some questions here and there to make sure my understanding is correct. But, a lot of the comments are just small nits.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -127,15 +130,20 @@ static CLinker getInstance() {
      * @return the downcall method handle.
-     * @throws IllegalArgumentException in the case of a method type and function descriptor mismatch.
+     * @throws IllegalArgumentException in the case of a method type and function descriptor mismatch,
+     * or if {@code type} has a prefix carrier of type {@link SegmentAllocator} but the return descriptor
```
I think this is a typo when I rewrote the javadoc. The SegmentAllocator parameter is inferred, and is not required in the method type. So I think it already works as you expect but the javadoc is wrong.

JornVernee commented on test/jdk/java/foreign/TestResourceScope.java:
```diff
@@ -0,0 +1,247 @@
+                    scopeRef.get().addOnClose(() -> {
+                        acc.addAndGet(delta);
+                    });
+                } catch (IllegalStateException ex) {
+                    // already closed - we need to call cleanup manually
+                    acc.addAndGet(delta);
+                }
```
Okay, that seems fair.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAddress.java:
```diff
@@ -91,72 +89,93 @@ default MemoryAddress address() {
+     * This method is equivalent to the following code:
      * <pre>{@code
-    asSegmentRestricted(byteSize, null, null);
+    asSegmentRestricted(byteSize, null, ResourceScope.ofShared(Cleaner.create()));
```
yep -good catch

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAddress.java:
```diff
@@ -91,72 +89,93 @@ default MemoryAddress address() {
-     * (see {@link MemorySegment#ALL_ACCESS}), and its confinement thread is the current thread (see {@link Thread#currentThread()}).
-     * Moreover, the returned segment will keep a strong reference to the supplied attachment object (if any), which can
-     * be useful in cases where the lifecycle of the segment is dependent on that of some other external resource.
+     * Returns a new confined native memory segment with given size, and whose base address is this address. This method
```
yep

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java:
```diff
@@ -593,11 +376,10 @@ default MemorySegment asSlice(MemoryAddress newBase) {
+     * after the scope associated with this segment has been closed (see {@link ResourceScope#close()}, will throw an {@link IllegalStateException}.
      * <p>
-     * If this segment is <em>shared</em>, calling certain I/O operations on the resulting buffer might result in
+     * If this segment is associated with a shared scope, calling certain I/O operations on the resulting buffer might result in
      * an unspecified exception being thrown. Examples of such problematic operations are {@link FileChannel#read(ByteBuffer)},
```
For now this is still the case - but Chris has put together a patch which helps removing this restriction - so this will go away.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java:
```diff
@@ -0,0 +1,262 @@
+ *
+ * Resource scopes can be associated with a {@link Cleaner} instance (see {@link #ofConfined(Cleaner)}) - we call these
+ * resource scopes <em>managed</em> resource scopes. A managed resource scope is closed automatically once the scope instance
+ * becomes <em>unreachable</em>.
```
good idea!

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java:
```diff
@@ -0,0 +1,262 @@
+     * An optional attachment can be associated with the resulting scope.
+     * @param attachment an attachment object which is kept alive by the returned resource scope (can be {@code null}).
+     * @param cleaner the cleaner to be associated with the returned scope. Can be {@code null}.
+     * @return a new confined scope, managed by {@code cleaner}; the resulting scope is closeable if {@code closeable == true}.
```
whoops!

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:
```diff
@@ -76,51 +74,44 @@ public static LibraryLookup load(String path) {
+                loadedLibraries.remove(library);
+            }));
+            return new WeakReference<>(s);
+        });
+        return new LibraryLookupImpl(library, scopeRef.get());
```
ok

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java:
```diff
@@ -98,21 +98,28 @@ public static MemorySegment makeNativeSegment(long bytesSize, long alignmentByte
+        scope.checkValidStateSlow();
+        AbstractMemorySegmentImpl segment = new NativeMemorySegmentImpl(min.toRawLongValue(), bytesSize, defaultAccessModes(bytesSize), scope);
+        if (cleanupAction != null) {
+            scope.addOnClose(cleanupAction);
```
This is a bit controversial: should the cleanup action be called if there's a race and, by the time we create the unsafe segment, the scope is already closed? Normally, we call `addOrCleanupIfFail` because we don't want e.g. `MemorySegment::allocateNative` to leave resource around if the scope happens to be closed - but in this case? I think I stick by the current decision - otherwise we're essentially making a guarantee that the cleanup action provided by the user will always be called, which I'm not sure I wanna make.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java:
```diff
@@ -0,0 +1,136 @@
+
+        public abstract void cleanup();
+
+        final static ResourceCleanup DUMMY_CLEANUP = new ResourceCleanup() {
```
yeah - I agree that it's used as a way to signal "list is closed".

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java:
```diff
@@ -0,0 +1,136 @@
+        cleanup(); // cleaner interop
+    }
+
+    final void cleanup(ResourceCleanup first) {
```
no strong preference

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java:
```diff
@@ -0,0 +1,136 @@
+                ResourceCleanup prev = null;
+                while (true) {
+                    prev = (ResourceCleanup) FST.getAcquire(this);
+                    // no need to check for DUMMY, since only one thread can get here!
```
Yeah - I started off this way, then moved off to the current organization as MemoryScope was getting biggie - but yes, we should do something better here

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java:
```diff
@@ -0,0 +1,136 @@
+                    }
+                }
+                cleanup(prev);
+            }
```
ok, good idea

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ArenaAllocator.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public synchronized MemorySegment allocate(long bytesSize, long bytesAlignment) {
```
Ok that sounds like a good idea.

mcimadamore replied:
Thinking more about this... thread local works fine for unbounded, not so much for bounded. I think this issue seems isolated enough that would probably better be addressed in a separate change.

mcimadamore replied:
I think I managed to come up with a way to solve this. As anticipated, shared unbounded allocator is now using a thread local, and just delegating allocation to the thread's confined allocator. I then removed all synchronization, except in the case of shared/bounded; I think that's acceptable (I don't think a bounded arena will really be used in a shared context).

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java:
```diff
@@ -0,0 +1,136 @@
+        }
+
+        void cleanup() {
+            if (fst != ResourceCleanup.DUMMY_CLEANUP) {
```
Probably - but, to be clear - at this point (and I need to add doc on this), we are only interested about add vs. close races - not close vs. close. So, the only "bad" thing that could happen is that some other thread adds to this list while we're closing it. So I guess in this case I omitted the read acquire because the only case where you could read a DUMMY_CLEANUP is if somebody has already stored that in there (otherwise you'll see something else, and move on). But no harm in making it more robust.

Another way we could go would be to manage races in the MemoryScope impl. That is, in a shared memory scope we could have some intermediate state (e.g. ADDING) which we go through and which is mutually exclusive w.r.t. CLOSE. I didn't want to go down that path since then in ResourceList I'd have to deal with add vs. add races anyway.

JornVernee replied:
Ok, that makes sense. I don't think an ADDING state is needed, but would be nice if you could comment this part.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java:
```diff
@@ -287,21 +351,25 @@ void justClose() {
-            this.cleanupAction = cleanupAction;
+            @Override
+            public void close() {
+                if (released.compareAndSet(false, true)) {
```
yep - we only need to check the thread. If the handle is not released then by definition the scope is alive.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java:
```diff
@@ -216,16 +219,103 @@
+        /**
+         * Create a binding context from given native scope.
+         */
+        public static Context ofBoundedAllocator(ResourceScope scope, long size) {
```
It's possible that there could be some simplification here - the code went back and forth a bit.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -258,7 +263,7 @@ public static void checkFunctionTypes(MethodType mt, FunctionDescriptor cDesc, l
         }
 
-        throw new IllegalArgumentException("No type for size: " + size + " isFloat=" + useFloat);
+        throw new IllegalArgumentException("Size too large: " + size);
```
ok

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -357,17 +371,29 @@ static MethodHandle wrapWithAllocator(MethodHandle specializedHandle,
+            scopeFilter = MH_SCOPE_TO_BINDING_CONTEXT;
+        } else {
+            // this path is probably never used now, since ProgrammableInvoker never calls this routine with bufferCopySize == 0
+            scopeFilter = insertArguments(identity(Binding.Context.class), 0, Binding.Context.DUMMY);
```
true!

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -440,6 +466,18 @@ public static MethodType convertVaListCarriers(MethodType mt, Class<?> carrier)
         return methodType(mt.returnType(), params);
     }
 
+    public static MethodHandle adaptDowncall(MethodType type, FunctionDescriptor descriptor,
+                                             MethodHandle MH_unboxVaList, Function<MethodType, MethodHandle> downcallFactory) {
+        MethodType sigMethodType = type;
+        MethodHandle handle = downcallFactory.apply(sigMethodType);
+        if (!sigMethodType.returnType().equals(MemorySegment.class)) {
+            // not returning segment, just insert default allocator
+            handle = MethodHandles.insertArguments(handle, 1, Binding.Context.DEFAULT.allocator());
+        }
+        handle = SharedUtils.unboxVaLists(type, handle, MH_unboxVaList);
+        return handle;
+    }
```
This method started off very complex and now ended up much simpler - we can probably revert the various call sites and just add a shared functionality to inject the allocator.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/AArch64VaList.java:
```diff
@@ -314,8 +309,8 @@ private Object read(Class<?> carrier, MemoryLayout layout, SharedUtils.Allocator
 
-                    try (MemorySegment slice = handoffIfNeeded(ptr
-                            .asSegmentRestricted(layout.byteSize()), segment.ownerThread())) {
+                    try (ResourceScope scope = ResourceScope.ofConfined()) {
```
not sure I get the suggestion, but when looking at this my feeling was that this method was just copying some data off to some other segment which is then allocated using the provided allocator. I believe what is missing here is a check that the guy calling read() can in fact do that (e.g. if the valist is confined, is the thread the same?) - but I'm not sure that, other than the check, we need something else?

JornVernee replied:
Yeah, AFAIK the only thing we need to check here is that the caller can actually read the segment. In fact, I don't think we need to create a scope here at all, since we don't return the created scope, we can just use the global scope I think.

mcimadamore replied:
I think the right thing to do here is to use the VaList scope, not the global scope. That way, when reading from the va list segment we will be protected from read after close issues (since the slice we created would be backed by the same scope as the va list).

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/windows/WinVaList.java:
```diff
@@ -148,57 +147,39 @@ public void skip(MemoryLayout... layouts) {
-        public Builder(SharedUtils.Allocator allocator) {
-            this.allocator = allocator;
+        public Builder(ResourceScope scope) {
+            ((MemoryScope)scope).checkValidStateSlow();
```
because in the other cases it happens implicitly via allocation

JornVernee replied:
Ah, ok

mcimadamore commented on test/jdk/java/foreign/TestAddressHandle.java:
```diff
@@ -62,7 +62,8 @@
         VarHandle longHandle = MemoryLayouts.JAVA_LONG.varHandle(long.class);
-        try (MemorySegment segment = MemorySegment.allocateNative(8)) {
+        try (ResourceScope scope = ResourceScope.ofConfined()) {
+            MemorySegment segment = MemorySegment.allocateNative(8, 8, scope);
```
probably was written before the overload was added

mcimadamore commented on test/jdk/java/foreign/TestNativeScope.java:
```diff
@@ -44,16 +40,12 @@
-import static jdk.incubator.foreign.MemorySegment.HANDOFF;
 import static org.testng.Assert.*;
 
 public class TestNativeScope {
```
yes

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -512,95 +568,60 @@ static void freeMemoryRestricted(MemoryAddress addr) {
-         * @throws IllegalStateException if the C {@code va_list} associated with this instance is no longer valid
-         * (see {@link #close()}).
-         */
-        VaList copy(NativeScope scope);
```
I double checked - I think the problem is that some implementations need to allocate, some don't. On SysV and Aarch, it seems to make sense to pass a scope, since you can actually create a new VaList, with a completely different scope. But on Windows? It seems like we're just creating another instance, backed by the same segment - so it's not clear how the scope would be used here. In other words, on Windows, the specified scope will say nothing about the lifecycle of the copy.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -644,17 +684,17 @@ static VaList ofAddressRestricted(MemoryAddress address) {
          * <p>
          * If this method needs to allocate native memory for the va list, it will use
-         * the given {@code NativeScope} to do so.
+         * the given {@code NativeAllocator} to do so.
```
I don't seem to have this in my code - probably forgot to push all my javadoc changes

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/LibraryLookup.java:
```diff
@@ -46,7 +46,7 @@
  * In cases where a client wants to create a memory segment out of a lookup symbol, the client might want to attach the
  * lookup symbol to the newly created segment, so that the symbol will be kept reachable as long as the memory segment
- * is reachable; this can be achieved by creating the segment using the {@link MemoryAddress#asSegmentRestricted(long, Runnable, Object)}
+ * is reachable; this can be achieved by creating the segment using the {@link MemoryAddress#asSegmentRestricted(long, Runnable, ResourceScope)}.
```
Same here

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -360,6 +357,7 @@ static MethodHandle swapArguments(MethodHandle mh, int firstArg, int secondArg)
     static MethodHandle wrapWithAllocator(MethodHandle specializedHandle,
                                           int allocatorPos, long bufferCopySize,
                                           boolean upcall) {
+        System.err.println(specializedHandle.type().descriptorString());
```
Don't forget to remove this.

mcimadamore commented:
I think I've address all of @JornVernee  comments - thanks again for the in-depth review!

JornVernee reviewed (approved):
Looks really good!

mcimadamore commented:
Here's a link to the latest javadoc:
http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v4/javadoc/jdk/incubator/foreign/package-summary.html

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java:
```diff
@@ -0,0 +1,262 @@
+ * associated with this resource scope. Any attempt to perform resource access from a thread other than the
+ * owner thread will result in a runtime failure.
+ * <p>
+ * Shared resource scopes (see {@link #ofShared()}), support strong thread-confinement guarantees. A shared resource scope
```
Noticed from the generated javadoc; is this a typo? "support string thread-confinement guarantees".

PaulSandoz reviewed (commented):
Very nice. A good improvement, with modest disruption (based on changes to the tests). Just some minor comments.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ArenaAllocator.java:
```diff
@@ -0,0 +1,109 @@
+
+import java.util.OptionalLong;
+
+public class ArenaAllocator implements SegmentAllocator {
```
I believe here there is some effort towards making everything thread-safe - in the sense that, based on the scope provided by the client (confined or shared) a different kind of allocator impl will be returned to the user. A shared bounded arena has a synchronized allocation method - while a shared unbound segment uses multiple underlying thread-local allocators.

Perhaps you refer to the fact that we don't check (in confined cases) that we are in the same thread as the scope's thread _before_ responding to an allocation request?

mcimadamore replied:
I've added confinement check in latest iteration

PaulSandoz replied:
Ok, i realized that i missed the `synchronized` on `BoundedSharedArenaAllocator.allocate`.

Perhaps we need to say in the `SegmentAllocator` something like:

If the allocator's resource scope is shared, then allocation is thread-safe and may be performed concurrently. If the resource scope is confined then allocation is confined to the owning thread of the allocator's resource scope.

And for the case of `SegmentAllocator.of(MemorySegment segment)` the allocator's resource scope is the scope of the given segment. 

Re-reading the specification i am confused by "The segment returned by this method is associated with a segment which cannot be closed".

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAddress.java:
```diff
@@ -91,72 +89,92 @@ default MemoryAddress address() {
      * @param bytesSize the desired size.
-     * @return a new confined native memory segment with given base address and size.
+     * @return a new native memory segment with given base address and size.
      * @throws IllegalArgumentException if {@code bytesSize <= 0}.
```
should be <= 0 everywhere (including for alignment) - good catch

mcimadamore commented:
Latest javadoc here:
http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v5/javadoc/jdk/incubator/foreign/package-summary.html

mcimadamore commented:
Latest javadoc here:
http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v6/javadoc/jdk/incubator/foreign/package-summary.html

mcimadamore commented:
A possible, improved naming for the two segment allocators:

SegmentAllocator.of(MemorySegment) -> SegmentAllocator.slicing(MemorySegment)
SegmentAllocator.of(ResourceScope) -> SegmentAllocator.scoped(ResourceScope)

What do you think?

PaulSandoz commented:
> A possible, improved naming for the two segment allocators:
> 
> SegmentAllocator.of(MemorySegment) -> SegmentAllocator.slicing(MemorySegment)
> SegmentAllocator.of(ResourceScope) -> SegmentAllocator.scoped(ResourceScope)
> 
> What do you think?

Yes, that's better. I find it hard to come up with a better concise name for the former, allocation returns a sliced prefix of the given segment, thus allocations are aliased, which is fundamentally different to all the other cases.

mcimadamore commented:
> > A possible, improved naming for the two segment allocators:
> > SegmentAllocator.of(MemorySegment) -> SegmentAllocator.slicing(MemorySegment)
> > SegmentAllocator.of(ResourceScope) -> SegmentAllocator.scoped(ResourceScope)
> > What do you think?
> 
> Yes, that's better. I find it hard to come up with a better concise name for the former, allocation returns a sliced prefix of the given segment, thus allocations are aliased, which is fundamentally different to all the other cases.

One option which I still find kind of appealing, is to add _instance_ methods to connect the various APIs. E.g. instead:

 SegmentAllocator.of(scope) -> scope.toAllocator();
 SegmentAllocator.of(segment) -> segment.toSegment();
 
 I think these conversions are great in certain contexts: think of a client slicing a segment and wanting to put the result of a native call in a certain place:

```
MemorySegment segment = ...
MemorySegment out = segment.asSlice(...);
linkerHandle.invokeExact(out.toAllocator(), ...);
```

Similarly, if you have an resource scope and simply want an allocator that returns segments attached to that scope:

```
try (ResourceScope scope = ...) {
   linkerHandle.invokeExact(scope.toAllocator(), ...);
}
```

I find both more concise that their static variants, but also (crucially) easier to explain. What we lose if we go down that path is "independence" between the various abstractions - in the sense that a ResourceScope is no longer Panama-neutral.

mcimadamore commented:
I've been doing some validation, e.g. port jextract to the new API:

https://github.com/openjdk/panama-foreign/compare/foreign-jextract...mcimadamore:jextract+resourceScope?expand=1

I think overall the results is neat. I liked how the new API doesn't _force_ me to have a scope for everything (there are many cases where we need to create a segment on the fly, as a step to be able to call a function once) - simple cases can just rely on cleaner and the GC, thus simplifying the code. At the same time, complex cases where lifecycle of multiple resources is linked, can be expressed way more clearly, since the user is now in control of the previously hidden "scope" component.

This patch adds overloads to native function wrappers returning structs, so that users can pass a NativeScope if they wish allocation to occur in that scope.

To minimize disruption, jextract is generating NativeScope in the target package; I think _some_ abstraction linking together ResourceScope and SegmentAllocator would be nice to have for the native use case, but we're unsure as to what the best way to expose this is.

One option would be to have a simple subtype of SegmentAllocator called ScopedSegmentAllocator - which expresses the notion of "I'm allocator that has its own scope". I think this will also be useful for integration with Jim's QBA allocator.

Another option would be to associate an _optional_ scope to every SegmentAllocator. That is, add a default method to SegmentAllocator:

default Optional<ResourceScope> allocatorScope() { returns Optional.empty(); }

Honestly, now that I look back at this, it doesn't seem as bad as I originally thought - most of the current allocators actually already have a natural scope - the only exception being malloc() - which doesn't have a scope, since all the returned segments are independent. Thoughts?

PaulSandoz commented:
> One option which I still find kind of appealing, is to add _instance_ methods to connect the various APIs. E.g. instead:
> 
> SegmentAllocator.of(scope) -> scope.toAllocator();
> SegmentAllocator.of(segment) -> segment.toSegment();
>

I think that works very well for `SegmentAllocator.of(segment)`, and implementation it could be quite efficient (same concrete class if need be).

Less sure about `SegmentAllocator.of(scope)` -> ResourceScope.toAllocator()`, FWIW I noticed this is a subset of `malloc`:

ResourceScope scope = ...
SegmentAllocator fixedScopeAllocator = SegmentAllocator.malloc(() -> scope);

There might also be some interconnections with `ScopedSegmentAllocator`? Seems like we either want to compose through inheritance or have some bi-direction between the two abstractions.

mcimadamore commented:
> Less sure about `SegmentAllocator.of(scope)` -> `ResourceScope.toAllocator()`, FWIW I noticed this is a subset of `malloc`:
> 
> ResourceScope scope = ...
> SegmentAllocator fixedScopeAllocator = SegmentAllocator.malloc(() -> scope);
> 
> There might also be some interconnections with `ScopedSegmentAllocator`? Seems like we either want to compose through inheritance or have some bi-direction between the two abstractions.

Good point - if we had a ScopedSegmentAllocator abstraction, I believe this method would be essentially useless (if clients need both an allocator and a scope, well they can just allocate a ScopedSegmentAllocator).

mcimadamore commented:
> > Less sure about `SegmentAllocator.of(scope)` -> `ResourceScope.toAllocator()`, FWIW I noticed this is a subset of `malloc`:
> > ResourceScope scope = ...
> > SegmentAllocator fixedScopeAllocator = SegmentAllocator.malloc(() -> scope);
> > There might also be some interconnections with `ScopedSegmentAllocator`? Seems like we either want to compose through inheritance or have some bi-direction between the two abstractions.
> 
> Good point - if we had a ScopedSegmentAllocator abstraction, I believe creating an allocator from a scope would not be as important (if clients need both an allocator and a scope, well they can just allocate a ScopedSegmentAllocator).

I made an attempt at this - captured in the following javadoc:

http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v6_exp/

There is a new ScopedMemoryAllocator - and some of the existing factory now return that. Note that now the difference between SegmentAllocator.malloc and SegmentAllocator.scoped is cleared, as only the latter gives you a ScopedSegmentAllocator.

All the arena allocators are also ScopedSegmentAllocators.

The recycling allocator has been moved off to the side in MemorySegment::toAllocator.

Is this an improvement? While more verbose than the old native scope, I think there's a lot more composition going on here, as now users can select full spectrum of { shared, confined } x { bounded vs. unbounded allocation } x { explicit vs. implicit close } - the former NativeScope used to be a point (although common) in this space: confined x { bounded vs. unbounded allocation } x explicit.

mcimadamore commented:
Small observation: I realized that _all_ segment allocators can be associated with a scope. For allocators in which the scope is not important, we can just return the global scope.

In other words, associating a scope to an allocator is mostly harmless - either the allocator does have a lifecycle (and that is managed by the scope), or the allocator does not have one (in which case a neutral scope like global scope can be used).

This might address the problem of having a single entity which is connected to both an allocator and a scope - but it doesn't solve the usability problem of the try-with-resources. Unless we make SegmentAlocator AutoCloseable. But, if we make SegmentAllocator auto-closeable on the basis that an allocator has-a scope - then, on that basis it seems that even MemorySegment, VaList, ... should join for the ride? This seems like a slippery slope.

mcimadamore commented:
To bring this discussion closer to actual use cases - the thing that worries me is examples like these:

```
          try (var url = toCString(urlStr)) {
               curl_easy_setopt(curl, CURLOPT_URL(), url.address());
               int res = curl_easy_perform(curl);
               if (res != CURLE_OK()) {
                   String error = toJavaStringRestricted(curl_easy_strerror(res));
                   System.out.println("Curl error: " + error);
                   curl_easy_cleanup(curl);
               }
           }
```
(this can be found in the list of jextract samples, in [1]).

This code just needs a native string (there is a similar case in the clang implementation used by jextract). Now, there are many ways to rewrite that code (for simplicity I'll just focus on the first two lines):

1. remove the try with resource - and just use "default" scope (which is managed by the GC):

```
var url = toCString(urlStr);
curl_easy_setopt(curl, CURLOPT_URL(), url.address());
```
But there's a sneaky problem here: the function uses the address projection of the string segment - meaning that, at least in principle, the string segment can become unreachable (and hence freed) _before_ the native function is executed. This seems bad - and to avoid that, explicit resource deallocation seems preferrable (unless we want to start playing with reachability fences).

2. Use a native scope - after all this is jextracted code:

```
try (NativeScope scope = NativeScope.unboundedScope())) {
    var url = toCString(urlStr, scope);
    curl_easy_setopt(curl, CURLOPT_URL(), url.address());
    ...
}
```
This works as expected. There are two issues here: non-jextracted code cannot do this (as there's no NativeScope in the API); and, arguably, using an unbounded native scope just to allocate a single string seems overkill. That will probably allocate more memory than required, to name one.

3. Use ResourceScope + SegmentAllocator

```
try (ResourceScope scope = ResourceScope.ofConfined()) {
    var url = toCString(urlStr, SegmentAllocator.scoped(scope));
    curl_easy_setopt(curl, CURLOPT_URL(), url.address());
    ...
}
```
This works, and has the same semantics as the code that it's replacing.

Now maybe (3) is the best we can do - but I can't help but feeling that code like this will be common (e.g. converting a string on the fly and passing it to some native library) and having to create a scope _and_ an allocator to be able to do that (so that the right semantics occurs) seems unfortunate.

[1] - https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_jextract.md

mcimadamore commented:
I've updated the branch. The main change is that now methods in CLinker have an overload accepting a ResourceScope - e.g. whenever we took an allocator, there is also a corresponding overload that takes only a scope. This way the user has a nice and gentle progression through the API complexity:

1. no scopes - everything is GC-managed (a la byte buffer)
2. deterministic closing with scopes - user creates a bunch of ResourceScope and all memory allocation is associated with such scopes
3. custom allocators - the user is not happy with default allocation mechanism - so, in addition to create a scope, a segment allocator will also be created.

While (1) and (2) will cover most of the "one-liner" examples, more realistic tests, such as those found in jextract will probably go for (3) [as they do today].

From here, we have a couple of API extensions that we can evaluate and add later (they can be added in a compatible fashion):

a. Make all entities that have a scope should also have a close() method- this means that, instead of scope().close(), users can say close(), and use a TWR
b. consider if the above entities should also be AutoCloseable
c. Consider whether all segment allocators should be associated with a scope. We could add a default method returning ResourceScope.globalScope so that compatibility is not affected.
d. if we do (c) maybe we could add more factories overloads e.g. an arenaBounded which takes no scope, but creates the scope in a single shot.

Note that in the latest change I've converted StdLib test not to use NativeScope. I think the result is uniform and pleasing, and I didn't find myself missing it.

I've also added some ResourceScope overloads to jextract:

https://github.com/mcimadamore/panama-foreign/tree/jextract%2BresourceScope

And tried to convert my pet OpenGL application not to use NativeScope and again, it's all pretty readable.

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -323,13 +321,30 @@ static MemorySegment toCString(String str, Charset charset) {
      *
      * @param str the Java string to be converted into a C string.
-     * @param scope the scope to be used for the native segment allocation.
+     * @param allocator the scope to be used for the native segment allocation.
```
```suggestion
     * @param allocator the allocator to be used for the native segment allocation.
```

See also use of scope above at line 315. Similarly for other methods.

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -343,14 +358,32 @@ static MemorySegment toCString(String str, SegmentAllocator scope) {
      * @param str the Java string to be converted into a C string.
      * @param charset The {@link java.nio.charset.Charset} to be used to compute the contents of the C string.
-     * @param scope the scope to be used for the native segment allocation.
+     * @param allocator the scope to be used for the native segment allocation.
```
```suggestion
     * @param allocator the allocator to be used for the native segment allocation.
```

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -343,14 +358,32 @@ static MemorySegment toCString(String str, SegmentAllocator scope) {
+
+    /**
+     * Converts a Java string into a null-terminated C string, using the given {@link java.nio.charset.Charset charset},
+     * storing the result into a new native memory segment native memory segment associated with the provided resource scope.
```
```suggestion
     * storing the result into a native memory segment associated with the provided resource scope.
```

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -573,16 +606,30 @@ static void freeMemoryRestricted(MemoryAddress addr) {
+         * The memory segment returned by this method will be allocated using the given {@link SegmentAllocator}.
+         *
+         * @param layout the layout of the value
+         * @param allocator the scope to be used for the native segment allocation
```
```suggestion
         * @param allocator the allocator to be used for the native segment allocation
```

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java:
```diff
@@ -181,6 +185,17 @@ protected Object clone() throws CloneNotSupportedException {
+     * Return a scoped allocator; the instance returned by this method is lazily created and then shared
+     * upon subsequent requests.
+     */
+    public SegmentAllocator allocator() {
```
Potentially `MemoryScope` could implement `SegmentAllocator` then there is no need for this method?

And then on `SegmentAllocator`:
```
    static SegmentAllocator scoped(ResourceScope scope) {
        return (MemoryScope) Objects.requireNonNull(scope);
    }
```

mcimadamore replied:
Yes - I guess I didn't want to blur the two concepts too much - if we go down that path I think users will be able to cast a ResourceScope into a SegmentAllocator?

PaulSandoz replied:
They could, but they should not since it's an implementation detail. (FWIW I don't think there is any impact to the integrity of the implementation.)

PaulSandoz reviewed (approved):
This is looking very good.

mcimadamore commented:
Addressed latest comments. I also added two new static methods, one to obtain a new "default scope" (that is, the implicit scope that is created when you call e.g. MemorySegment::allocateNative), and also the "default allocator" - e.g. an allocator that returns segments based on their own default scopes.

javadoc here:
http://cr.openjdk.java.net/~mcimadamore/panama/resourceScope-javadoc_v7/javadoc/jdk/incubator/foreign/package-summary.html

JornVernee reviewed (approved):
FWIW, all the changes still look good to me. I like the new names for the allocator factory methods and that overloads were added to CLinker that take a `ResourceScope` as well.

mcimadamore commented:
/integrate
